From cf0fb3a2d093cacea15231f268ae9959079b0428 Mon Sep 17 00:00:00 2001
From: Miroslav Grepl <mgrepl@redhat.com>
Date: Fri, 20 Jun 2014 10:39:06 +0200
Subject: [PATCH] Fix fixfiles to return zero value when no error is
 encountered

---
 policycoreutils/scripts/fixfiles | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/policycoreutils/scripts/fixfiles b/policycoreutils/scripts/fixfiles
index f333da3..f82f676 100755
--- a/policycoreutils/scripts/fixfiles
+++ b/policycoreutils/scripts/fixfiles
@@ -175,14 +175,14 @@ then
 /bin/cp $TEMPFCFILE $FC &>/dev/null || exit
 fi
 ${SETFILES} $exclude_dirs -q ${SYSLOGFLAG} ${FORCEFLAG} $* ${FC} ${FILESYSTEMS} 2>&1 | cat >> $LOGFILE
-if [ -e $TEMPFCBACKUP ];then
+if [ -n "${exclude_dirs}" ];then
     /bin/cp -p ${TEMPFCBACKUP} ${FC} &>/dev/null || exit
     rm -rf ${TEMPFCBACKUP}
 fi
 rm -rf /tmp/gconfd-* /tmp/pulse-* /tmp/orbit-* $TEMPFCFILE
 find /tmp \( -context "*:file_t*" -o -context "*:unlabeled_t*" \) -exec chcon -t tmp_t {} \;
 find /var/tmp \( -context "*:file_t*" -o -context "*:unlabeled_t*" \) -exec chcon -t tmp_t {} \;
-exit $?
+exit 0
 }
 
 fullrelabel() {
-- 
2.0.0

