From 2ccf508ff5bc439f915fe033669a4c2d7196687d Mon Sep 17 00:00:00 2001
From: Miroslav Grepl <mgrepl@redhat.com>
Date: Mon, 23 Jun 2014 09:55:44 +0200
Subject: [PATCH 1/5] Make sure file equivalance target and source do not end
 with a /

---
 policycoreutils/semanage/seobject.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/policycoreutils/semanage/seobject.py b/policycoreutils/semanage/seobject.py
index c4f9de3..191ca31 100644
--- a/policycoreutils/semanage/seobject.py
+++ b/policycoreutils/semanage/seobject.py
@@ -1686,6 +1686,10 @@ class fcontextRecords(semanageRecords):
 
         def add_equal(self, target, substitute):
                 self.begin()
+                if target != "/" and target[-1] == "/":
+                       raise ValueError(_("Target %s is not valid. Target is not allowed to end with '/'") % target )
+                if substitute != "/" and substitute[-1] == "/":
+                       raise ValueError(_("Substiture %s is not valid. Substitute is not allowed to end with '/'") % substitute )
                 if target in self.equiv.keys():
                        raise ValueError(_("Equivalence class for %s already exists") % target)
                 self.validate(target)
-- 
2.0.0

