From 40288f38e089be118f889ea9da3932be64098f67 Mon Sep 17 00:00:00 2001
From: Miroslav Grepl <mgrepl@redhat.com>
Date: Mon, 23 Jun 2014 10:34:37 +0200
Subject: [PATCH 3/5] Fix polgen.py to allow _ in a policy file.

---
 policycoreutils/gui/polgen.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/policycoreutils/gui/polgen.py b/policycoreutils/gui/polgen.py
index e4f1f0b..212c4b1 100644
--- a/policycoreutils/gui/polgen.py
+++ b/policycoreutils/gui/polgen.py
@@ -313,7 +313,7 @@ class policy:
 ( self.generate_sandbox_types, self.generate_sandbox_rules))
 		if name == "":
 			raise ValueError(_("You must enter a name for your confined process/user"))
-                if not name.isalnum():
+                if not re.match(r"^[a-zA-Z0-9-_]+$", name)::
                     raise ValueError(_("Name must be alpha numberic with no spaces. Consider using option \"-n MODULENAME\""))
 
 		if type == CGI:
-- 
2.0.0

