From 72f058beece5865f6687abab480f224fc8ccf24b Mon Sep 17 00:00:00 2001
From: Miroslav Grepl <mgrepl@redhat.com>
Date: Mon, 23 Jun 2014 13:34:59 +0200
Subject: [PATCH 1/2] Make -q and -d options mutually exclusive

---
 policycoreutils/setfiles/setfiles.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/policycoreutils/setfiles/setfiles.c b/policycoreutils/setfiles/setfiles.c
index 64c0859..77aa693 100644
--- a/policycoreutils/setfiles/setfiles.c
+++ b/policycoreutils/setfiles/setfiles.c
@@ -267,6 +267,10 @@ int main(int argc, char **argv)
 			input_filename = optarg;
 			break;			
 		case 'd':
+			if (r_opts.quiet) {
+				fprintf(stderr,"-q and -d options mutually exclusive.\n");
+				exit(1);
+			}
 			r_opts.debug = 1;
 			break;
 		case 'i':
@@ -297,6 +301,10 @@ int main(int argc, char **argv)
 			__fsetlocking(r_opts.outfile, FSETLOCKING_BYCALLER);
 			break;
 		case 'q':
+			if (r_opts.debug) {
+				fprintf(stderr,"-q and -d options mutually exclusive.\n");
+				exit(1);
+			}
 			r_opts.quiet = 1;
 			break;
 		case 'R':
-- 
2.0.0

