From 129783431831489a765be79ee3eff12088c342dc Mon Sep 17 00:00:00 2001
From: Miroslav Grepl <mgrepl@redhat.com>
Date: Thu, 3 Jul 2014 10:37:46 +0200
Subject: [PATCH] Check if all files exist for setfiles and restorecon

---
 policycoreutils/setfiles/restore.c | 17 ++++++++---------
 1 file changed, 8 insertions(+), 9 deletions(-)

diff --git a/policycoreutils/setfiles/restore.c b/policycoreutils/setfiles/restore.c
index c6ae5a2..6628598 100644
--- a/policycoreutils/setfiles/restore.c
+++ b/policycoreutils/setfiles/restore.c
@@ -433,19 +433,18 @@ int process_one_realpath(char *name, int recurse)
 			"Must call initialize first!");
 		return -1;
 	}
+    rc = lstat64(name, &sb);
+    if (rc < 0) {
+        if (r_opts->ignore_enoent && errno == ENOENT)
+            return 0;
+        fprintf(stderr, "%s:  lstat(%s) failed:  %s\n",
+            r_opts->progname, name, strerror(errno));
+        return -1;
+    }
 
 	if (!r_opts->expand_realpath) {
 		return process_one(name, recurse);
 	} else {
-		rc = lstat64(name, &sb);
-		if (rc < 0) {
-			if (r_opts->ignore_enoent && errno == ENOENT)
-				return 0;
-			fprintf(stderr, "%s:  lstat(%s) failed:  %s\n",
-				r_opts->progname, name,	strerror(errno));
-			return -1;
-		}
-
 		if (S_ISLNK(sb.st_mode)) {
 			char path[PATH_MAX + 1];
 
-- 
2.0.0

