From 07bad1cae226a7daf8b735af426f9dfa46c3c1e7 Mon Sep 17 00:00:00 2001
From: Miroslav Grepl <mgrepl@redhat.com>
Date: Thu, 3 Jul 2014 11:10:04 +0200
Subject: [PATCH] Fail properly on invalid options for restorecon/setfiles when
 using -R or -r

---
 policycoreutils/setfiles/setfiles.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/policycoreutils/setfiles/setfiles.c b/policycoreutils/setfiles/setfiles.c
index 77aa693..433cf19 100644
--- a/policycoreutils/setfiles/setfiles.c
+++ b/policycoreutils/setfiles/setfiles.c
@@ -145,6 +145,9 @@ int main(int argc, char **argv)
 	int recurse; /* Recursive descent. */
 	char *base;
 	int mass_relabel = 0, errors = 0;
+    const char *ropts = "e:f:hilno:pqrsvFRW0";
+    const char *sopts = "c:de:hilno:pqr:svFR:W";
+    const char *opts;
 	
 	memset(&r_opts, 0, sizeof(r_opts));
 
@@ -217,8 +220,13 @@ int main(int argc, char **argv)
 	/* This must happen before getopt. */
 	exclude_non_seclabel_mounts();
 
+    if (iamrestorecon)
+        opts = ropts;
+    else
+        opts = sopts;
+
 	/* Process any options. */
-	while ((opt = getopt(argc, argv, "c:de:f:ilnpqrsvo:FRW0")) > 0) {
+    while ((opt = getopt(argc, argv, opts)) > 0) {
 		switch (opt) {
 		case 'c':
 			{
@@ -324,7 +332,7 @@ int main(int argc, char **argv)
 					argv[0]);
 				exit(1);
 			}
-			set_rootpath(argv[optind++]);
+			set_rootpath(argv[optind]);
 			break;
 		case 's':
 			use_input_file = 1;
-- 
2.0.0

