From 3871118e81c94f7205e3f5ba73100a6e71df0691 Mon Sep 17 00:00:00 2001
From: Miroslav Grepl <mgrepl@redhat.com>
Date: Thu, 2 Oct 2014 12:25:01 +0200
Subject: [PATCH] Fix semanageRecords() to define load variable.

---
 policycoreutils/semanage/seobject.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/policycoreutils/semanage/seobject.py b/policycoreutils/semanage/seobject.py
index 5481182..0e1ef97 100644
--- a/policycoreutils/semanage/seobject.py
+++ b/policycoreutils/semanage/seobject.py
@@ -236,7 +236,7 @@ class semanageRecords:
         store = None
         def __init__(self, store):
                global handle
-
+               self.load =  True
                self.sh = self.get_handle(store)
 
 	       rc, localstore = selinux.selinux_getpolicytype()
-- 
2.1.0

