diff --git a/audit2allow/audit2allow b/audit2allow/audit2allow
index 2917c84..1ebe91f 100644
--- a/audit2allow/audit2allow
+++ b/audit2allow/audit2allow
@@ -29,6 +29,8 @@ import sepolgen.defaults as defaults
 import sepolgen.module as module
 from sepolgen.sepolgeni18n import _
 import selinux.audit2why as audit2why
+import locale
+locale.setlocale(locale.LC_ALL, '')
 
 class AuditToPolicy:
     VERSION = "%prog .1"
diff --git a/sepolgen-1.0.23/src/sepolgen/audit.py b/sepolgen-1.0.23/src/sepolgen/audit.py
index 46faff2..28130f4 100644
--- a/sepolgen-1.0.23/src/sepolgen/audit.py
+++ b/sepolgen-1.0.23/src/sepolgen/audit.py
@@ -37,9 +37,8 @@ def get_audit_boot_msgs():
     off=float(fd.read().split()[0])
     fd.close
     s = time.localtime(time.time() - off)
-    date = time.strftime("%D/%Y", s).split("/")
-    bootdate="%s/%s/%s" % (date[0], date[1], date[3])
-    boottime = time.strftime("%X", s)
+    bootdate = time.strftime("%x", s)
+    boottime = time.strftime("%T", s)
     output = subprocess.Popen(["/sbin/ausearch", "-m", "AVC,USER_AVC,MAC_POLICY_LOAD,DAEMON_START,SELINUX_ERR", "-ts", bootdate, boottime],
                               stdout=subprocess.PIPE).communicate()[0]
     return output
