diff --git a/scripts/fixfiles b/scripts/fixfiles
index f82f676..1d0e884 100755
--- a/scripts/fixfiles
+++ b/scripts/fixfiles
@@ -140,6 +140,9 @@ rpm -q --qf '[%{FILESTATES} %{FILENAMES}\n]' "$1" | grep '^0 ' | cut -f2- -d ' '
 # if called with -n will only check file context
 #
 restore () {
+OPTION=$1
+shift
+
 if [ ! -z "$PREFC" ]; then
     diff_filecontext $*
     exit $?
@@ -175,6 +178,9 @@ then
 /bin/cp $TEMPFCFILE $FC &>/dev/null || exit
 fi
 ${SETFILES} $exclude_dirs -q ${SYSLOGFLAG} ${FORCEFLAG} $* ${FC} ${FILESYSTEMS} 2>&1 | cat >> $LOGFILE
+if [ ${OPTION} != "Relabel" ]; then
+    return
+fi
 if [ -n "${exclude_dirs}" ];then
     /bin/cp -p ${TEMPFCBACKUP} ${FC} &>/dev/null || exit
     rm -rf ${TEMPFCBACKUP}
@@ -188,12 +194,12 @@ exit 0
 fullrelabel() {
     logit "Cleaning out /tmp"
     find /tmp/ -mindepth 1 -delete
-    restore
+    restore Relabel
 }
 
 relabel() {
     if [ ! -z "$RPMFILES" ]; then
-	restore 
+	restore Relabel
     fi
 
     if [ $fullFlag == 1  ]; then
@@ -210,7 +216,7 @@ relabel() {
     if [ "$answer" = y -o  "$answer" = Y ]; then 
 	fullrelabel
     else
-	restore
+	restore Relabel
     fi
 }
 
@@ -219,9 +225,9 @@ process() {
 # Make sure they specified one of the three valid commands
 #
 case "$1" in
-    restore) restore -p ;;
-    check) restore -n -v;;
-    verify) restore -n -o -;;
+    restore) restore Relabel -p ;;
+    check) restore Check -n -v;;
+    verify) restore Verify -n -o -;;
     relabel) relabel;;
     onboot)
 	touch /.autorelabel
