diff --git a/semanage/semanage b/semanage/semanage
index 41525f5..f402821 100644
--- a/semanage/semanage
+++ b/semanage/semanage
@@ -560,7 +560,10 @@ Object-specific Options (see above):
                              sys.stdout = open(output, 'w')
                       for i in manageditems:
                              print "%s -D" % i
-                             process_args([i, "-E"])
+                             if store == "":
+                                    process_args([i, "-E"])
+                             else:
+                                    process_args([i, "-E", "-S", store])
                       sys.exit(0)
 
                if input != None:
