From ce9c4000b0ab3d7de9c633c8f2084d87cd69d051 Mon Sep 17 00:00:00 2001
From: Petr Lautrbach <plautrba@redhat.com>
Date: Mon, 21 Dec 2015 08:37:42 +0100
Subject: [PATCH] fixfiles: check the SELinux status

---
 scripts/fixfiles | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/scripts/fixfiles b/scripts/fixfiles
index 1d0e884..0c7d7f1 100755
--- a/scripts/fixfiles
+++ b/scripts/fixfiles
@@ -140,6 +140,12 @@ rpm -q --qf '[%{FILESTATES} %{FILENAMES}\n]' "$1" | grep '^0 ' | cut -f2- -d ' '
 # if called with -n will only check file context
 #
 restore () {
+
+if ! selinuxenabled; then
+       echo $"SELinux is disabled."
+       exit 1
+fi
+
 OPTION=$1
 shift
 
@@ -198,6 +204,11 @@ fullrelabel() {
 }
 
 relabel() {
+    if ! selinuxenabled; then
+        echo $"SELinux is disabled."
+        exit 1
+    fi
+
     if [ ! -z "$RPMFILES" ]; then
 	restore Relabel
     fi
-- 
2.6.4

