From 6f5cf1a3715e7722e414b14ca0b7891869308f35 Mon Sep 17 00:00:00 2001
From: Petr Lautrbach <plautrba@redhat.com>
Date: Mon, 21 Dec 2015 08:58:30 +0100
Subject: [PATCH] semanage: check if store exists

---
 semanage/semanage | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/semanage/semanage b/semanage/semanage
index f402821..33f2bc8 100644
--- a/semanage/semanage
+++ b/semanage/semanage
@@ -21,6 +21,7 @@
 #
 #  
 import policycoreutils.default_encoding_utf8
+import os.path
 import sys, getopt, re
 import seobject
 import selinux
@@ -549,7 +550,10 @@ Object-specific Options (see above):
 
                for o, a in gopts:
                       if o == "-S" or o == '--store':
-                             store = a
+                             if os.path.exists("/etc/selinux/%s" % a):
+                                     store = a
+                             else:
+                                     raise ValueError ("Store %s cannot be accessed." % a)
                       if o == "-i" or o == '--input':
                              input = a
                       if o == "-o" or o == '--output':
-- 
2.6.4

