From c0dcef15fef7dbfb82adb8e740af0b2b7738b168 Mon Sep 17 00:00:00 2001
From: Petr Lautrbach <plautrba@redhat.com>
Date: Mon, 21 Dec 2015 11:18:12 +0100
Subject: [PATCH] semanage: don't skip reserver_port_t

---
 semanage/seobject.py | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/semanage/seobject.py b/semanage/seobject.py
index cf018d9..660b73e 100644
--- a/semanage/seobject.py
+++ b/semanage/seobject.py
@@ -1193,8 +1193,6 @@ class portRecords(semanageRecords):
 		for port in self.plist:
 			con = semanage_port_get_con(port)
 			ctype = semanage_context_get_type(con)
-			if ctype == "reserved_port_t":
-				continue
 			level = semanage_context_get_mls(con)
 			proto = semanage_port_get_proto(port)
 			proto_str = semanage_port_get_proto_str(proto)
@@ -1215,8 +1213,6 @@ class portRecords(semanageRecords):
 		for port in self.plist:
 			con = semanage_port_get_con(port)
 			ctype = semanage_context_get_type(con)
-			if ctype == "reserved_port_t":
-				continue
 			proto = semanage_port_get_proto(port)
 			proto_str = semanage_port_get_proto_str(proto)
 			low = semanage_port_get_low(port)
-- 
2.6.4

