From d9e447bfd3a6303fb8fd177acf1a4f4eb0f0e8bd Mon Sep 17 00:00:00 2001
From: Petr Lautrbach <plautrba@redhat.com>
Date: Wed, 20 Jan 2016 09:12:15 +0100
Subject: [PATCH] sandbox: Improve comments in sysconfig file

---
 sandbox/sandbox.conf | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/sandbox/sandbox.conf b/sandbox/sandbox.conf
index 7c35808..d33ed1a 100644
--- a/sandbox/sandbox.conf
+++ b/sandbox/sandbox.conf
@@ -1,4 +1,5 @@
-# Space separate list of homedirs
+# Space separate list of directories that should be made private
+# (not necessarily just home directories).
 HOMEDIRS="/home"
 # Control group configuration
 NAME=sandbox
-- 
1.8.3.1

