--- policycoreutils-2.0.83/sepolgen-1.0.23/src/sepolgen/yacc.py.fips	2012-04-18 08:31:30.773716465 +0200
+++ policycoreutils-2.0.83/sepolgen-1.0.23/src/sepolgen/yacc.py	2012-04-18 08:33:41.923719458 +0200
@@ -506,7 +506,7 @@
 
     Errorfunc    = None    # User defined error handler
 
-    Signature    = hashlib.md5()   # Digital signature of the grammar rules, precedence
+    Signature    = hashlib.sha256()   # Digital signature of the grammar rules, precedence
                                # and other information.  Used to determined when a
                                # parsing table needs to be regenerated.
 
