--- policycoreutils-2.0.83/semanage/seobject.py~	2012-04-02 15:34:04.485851339 +0200
+++ policycoreutils-2.0.83/semanage/seobject.py	2012-04-02 15:38:05.594856841 +0200
@@ -1631,14 +1631,6 @@
                        fd.close()
                 except IOError:
                        pass
-                try:
-                       fd = open(selinux.selinux_file_context_subs_dist_path(), "r")
-                       for i in fd.readlines():
-                              target, substitute = i.split()
-                              self.equiv_dist[target] = substitute
-                       fd.close()
-                except IOError:
-                       pass
 
         def commit(self):
                 if self.equal_ind:
