--- policycoreutils-2.0.83/run_init/run_init.c.old	2011-09-14 09:59:36.377127039 +0200
+++ policycoreutils-2.0.83/run_init/run_init.c	2011-09-14 10:04:38.863857266 +0200
@@ -414,10 +414,17 @@ int main(int argc, char *argv[])
 	 * execvp or using a exec(1) recycles pty's, and does not open a new
 	 * one. 
 	 */
+#ifdef USE_OPEN_INIT_PTY
 	if (execvp("/usr/sbin/open_init_pty", argv)) {
 		perror("execvp");
 		exit(-1);
 	}
+#else
+	if (execvp(argv[1], argv + 1)) {
+		perror("execvp");
+		exit(-1);
+	}
+#endif
 	return 0;
 
 }				/* main() */
