diff -Nur nsapolicycoreutils/load_policy/load_policy.c policycoreutils-2.0.83/load_policy/load_policy.c
--- nsapolicycoreutils/load_policy/load_policy.c	2010-06-16 13:04:12.000000000 +0100
+++ policycoreutils-2.0.83/load_policy/load_policy.c	2011-10-27 21:42:42.904837993 +0100
@@ -87,9 +87,14 @@
 		ret = selinux_mkload_policy(1);
 	}
 	if (ret < 0) {
-		fprintf(stderr, _("%s:  Can't load policy:  %s\n"),
-			argv[0], strerror(errno));
-		exit(2);
+		if ( errno ==  2 ) {
+			exit(2);
+		}
+		else {
+			fprintf(stderr, _("%s:  Can't load policy:  %s\n"),
+				argv[0], strerror(errno));
+			exit(2);
+		}
 	}
 	exit(0);
 }
